rust/connlib/shared/src/interface/linux.rs

```diff
@@ -0,0 +1,296 @@
+//! Virtual network interface
+
+use crate::{
+    callbacks::{Cidrv4, Cidrv6},
+    linux::{etc_resolv_conf, DnsControlMethod},
+};
+use anyhow::{anyhow, bail, Context as _, Result};
+use futures::TryStreamExt;
+use ip_network::{IpNetwork, Ipv4Network, Ipv6Network};
+use netlink_packet_route::route::{RouteProtocol, RouteScope};
+use netlink_packet_route::rule::RuleAction;
+use rtnetlink::{new_connection, Error::NetlinkError, Handle, RouteAddRequest, RuleAddRequest};
+use std::{
+    collections::HashSet,
+    net::{IpAddr, Ipv4Addr, Ipv6Addr},
+};
+
+// TODO: De-dupe before merging
+const FIREZONE_MARK: u32 = 0xfd002021;
+const IFACE_NAME: &str = "tun-firezone";
+const DEFAULT_MTU: u32 = 1280;
+const FILE_ALREADY_EXISTS: i32 = -17;
+const FIREZONE_TABLE: u32 = 0x2021_fd00;
+
+/// For lack of a better name
+pub struct InterfaceManager {
+    // This gets lazy-initialized when the interface is first configured
+    connection: Option<Connection>,
```

Comment:
Can we create the interface with `InterfaceManager::new`, we create it with no ip and then we just set it with `on_set_interface_config`.

Reply:
That would be very nice. These `Option` states that can effectively not be empty are always confusing.

Reply:
Yeah I thought there was some important reason for it like:

1. `InterfaceManager` is created
2. Something important happens in another piece of code.
3. The connection can be started

I thought the order of 2 and 3 was important. But I removed that `Option` and it still worked and I can't find anything that might be in Step 2